Public share middleware & controller

rullzer opened the pull request:
Fixes #5986

@juliushaertl as discussed.

Todo:
- [x] ~~integrate https://github.com/nextcloud/server/pull/9756~~
- [x] add since tags etc
- [x] Maybe nog change functions for publci sharing?
- [x] Move PublicFileShareController just to files_sharing
- [x] Move basic auth templates to core (so they are default by default as well)
- [x] add tests

Once merged:
- [x] Docs: https://github.com/nextcloud/documentation/issues/785

rullzer committed:
Add the new PublicShareController and PublicShareMiddleware

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Quick hack to make the files sharing app work with this properly

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

juliushaertl commented on lib/public/AppFramework/PublicShareController.php:
```diff
@@ -0,0 +1,142 @@
+	 * Show the authentication page
+	 * The form has to submit to the authenticate method route
+	 */
+	abstract public function showAuthenticate(string $token): TemplateResponse;
```
Shouldn't we have a general showAuthenticate implementation here as well, since most of the time, we will just render a passwort request form.

rullzer replied:
We could do that sure.. just server it from the core app I guess then.

juliushaertl commented:
@rullzer The IShareManager is just to be used by files, right? So calendar or other apps still need to implement their own sharing logic.

rullzer commented:
@juliushaertl that is correct.

Ok let me think how to make it more generic then.
And then maybe have a FilesPublicShareController for all the apps that want to share files that implements some extra stuff.

rullzer commented:
@juliushaertl ok abstracted away some more ;) This should work for all types of shares I guess

rullzer commented:
Ok so it probably should be even further split up. Some controllers just need to know you are properly authenticated (like preview endpoints). If not they will just 404... I'll look into that more next week.

rullzer committed:
Move public preview endpoint over

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

juliusknorr reviewed (commented):
@rullzer Looks good to me now in terms of separation :+1:


> ￼- Move basic auth templates to core (so they are default by default as well)

:+1:

rullzer commented on lib/public/AppFramework/FilesPublicShareController.php:
```diff
@@ -0,0 +1,79 @@
+use OCP\Share\IManager as ShareManager;
+use OCP\Share\IShare;
+
+abstract class FilesPublicShareController extends AuthPublicShareController {
```
In theory gallery could use it for example. BUt I agree. I'll move it. So the new code we add stays simple. If it turns out we duplicate it again in a lot (>4) places. Then we can think how to unify again.

rullzer replied:
Done

rullzer commented on apps/files_sharing/lib/Middleware/SharingCheckMiddleware.php:
```diff
@@ -102,12 +102,14 @@ public function beforeController($controller, $methodName) {
 			$token = $this->request->getParam('token');
 			$share = $this->shareManager->getShareByToken($token);
 			if ($share->getShareType() === \OCP\Share::SHARE_TYPE_LINK
 				&& !$this->isLinkSharingEnabled()) {
```
mmm good point.

Yeah probably makes sense to not allow any public shares when it is disabled.

rullzer replied:
Done

rullzer committed:
Move public auth page over

Now this is in core so the basics (that 99% of the app will want to
use) looks always the same.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Don't allow public share pages if link sharing is disabled

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Fix PublicPreviewControllerTests

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Fix ShareControllerTest

* Removed tests that are now handled by the middleware
* Updated tests

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Update SharingCheckMiddlewareTest

* They are handled now by the overal sharing public page middleware

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Add PublicShareMiddlewareTest

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Add PublicShareControllerTests

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Add AuthPublicShareControllerTest

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
Failing test is unrelated

rullzer committed:
Update token on regeneration for public login

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
Failing test is unrelated.

MorrisJobke commented:
* [ ] I created a public link share with password, switched to this branch and then refreshed the page. I typed in the password and then I get this internal server error:

```
Entfernte Adresse: 172.20.0.1
Anfragekennung: jOqhPAG8W3tWCHhbKINY
Typ: Symfony\Component\Routing\Exception\MissingMandatoryParametersException
Code: 0
Nachricht: Some mandatory parameters are missing ("token") to generate a URL for route "files_sharing.sharecontroller.showShare".
Datei: /var/www/html/3rdparty/symfony/routing/Generator/UrlGenerator.php
Zeile: 134

Trace

#0 /var/www/html/3rdparty/symfony/routing/Generator/UrlGenerator.php(119): Symfony\Component\Routing\Generator\UrlGenerator->doGenerate(Array, Array, Array, Array, Array, 'files_sharing.s...', 1, Array, Array)
#1 /var/www/html/lib/private/Route/Router.php(337): Symfony\Component\Routing\Generator\UrlGenerator->generate('files_sharing.s...', Array, 1)
#2 /var/www/html/lib/private/Route/CachingRouter.php(60): OC\Route\Router->generate('files_sharing.s...', Array, false)
#3 /var/www/html/lib/private/URLGenerator.php(79): OC\Route\CachingRouter->generate('files_sharing.s...', Array)
#4 /var/www/html/lib/public/AppFramework/AuthPublicShareController.php(188): OC\URLGenerator->linkToRoute('files_sharing.s...', Array)
#5 /var/www/html/lib/public/AppFramework/AuthPublicShareController.php(133): OCP\AppFramework\AuthPublicShareController->getRedirect()
#6 /var/www/html/lib/private/AppFramework/Http/Dispatcher.php(166): OCP\AppFramework\AuthPublicShareController->authenticate('abc123abc!')
#7 /var/www/html/lib/private/AppFramework/Http/Dispatcher.php(99): OC\AppFramework\Http\Dispatcher->executeController(Object(OCA\Files_Sharing\Controller\ShareController), 'authenticate')
#8 /var/www/html/lib/private/AppFramework/App.php(118): OC\AppFramework\Http\Dispatcher->dispatch(Object(OCA\Files_Sharing\Controller\ShareController), 'authenticate')
#9 /var/www/html/lib/public/AppFramework/App.php(140): OC\AppFramework\App::main('ShareController', 'authenticate', Object(OC\AppFramework\DependencyInjection\DIContainer))
#10 /var/www/html/core/routes.php(122): OCP\AppFramework\App->dispatch('ShareController', 'authenticate')
#11 [internal function]: OC\Route\Router->{closure}(Array)
#12 /var/www/html/lib/private/Route/Router.php(297): call_user_func(Object(Closure), Array)
#13 /var/www/html/lib/base.php(992): OC\Route\Router->match('/s/fb2YGarpyfLf...')
#14 /var/www/html/index.php(42): OC::handleRequest()
#15 {main}
```

danxuliu commented:
>  Failing test is unrelated.

I think that they are related to the changes in this pull request:
```
   Scenario: access a shared link protected by password with a valid password                   # /drone/src/github.com/nextcloud/server/tests/acceptance/features/app-files.feature:116
    Given I act as John                                                                        # ActorContext::iActAs()
    And I am logged in                                                                         # LoginPageContext::iAmLoggedIn()
    And I share the link for "welcome.txt" protected by the password "abcdef"                  # FilesAppContext::iShareTheLinkForProtectedByThePassword()
    And I write down the shared link                                                           # FilesAppContext::iWriteDownTheSharedLink()
    When I act as Jane                                                                         # ActorContext::iActAs()
    And I visit the shared link I wrote down                                                   # FilesSharingAppContext::iVisitTheSharedLinkIWroteDown()
    And I see that the current page is the Authenticate page for the shared link I wrote down  # FilesSharingAppContext::iSeeThatTheCurrentPageIsTheAuthenticatePageForTheSharedLinkIWroteDown()
      Failed asserting that two strings are equal.
      --- Expected
      +++ Actual
      @@ @@
      -'http://acceptance-app-files/index.php/s/RD7iZdksgbmZXiq/authenticate/preview'
      +'http://acceptance-app-files/index.php/s/RD7iZdksgbmZXiq/authenticate/showShare'
    And I authenticate with password "abcdef"                                                  # FilesSharingAppContext::iAuthenticateWithPassword()
    Then I see that the current page is the shared link I wrote down                           # FilesSharingAppContext::iSeeThatTheCurrentPageIsTheSharedLinkIWroteDown()
    And I see that the shared file preview shows the text "Welcome to your Nextcloud account!" # FilesSharingAppContext::iSeeThatTheSharedFilePreviewShowsTheText()

  Scenario: access a shared link protected by password with an invalid password                # /drone/src/github.com/nextcloud/server/tests/acceptance/features/app-files.feature:128
    Given I act as John                                                                        # ActorContext::iActAs()
    And I am logged in                                                                         # LoginPageContext::iAmLoggedIn()
    And I share the link for "welcome.txt" protected by the password "abcdef"                  # FilesAppContext::iShareTheLinkForProtectedByThePassword()
    And I write down the shared link                                                           # FilesAppContext::iWriteDownTheSharedLink()
    When I act as Jane                                                                         # ActorContext::iActAs()
    And I visit the shared link I wrote down                                                   # FilesSharingAppContext::iVisitTheSharedLinkIWroteDown()
    And I authenticate with password "fedcba"                                                  # FilesSharingAppContext::iAuthenticateWithPassword()
    Then I see that the current page is the Authenticate page for the shared link I wrote down # FilesSharingAppContext::iSeeThatTheCurrentPageIsTheAuthenticatePageForTheSharedLinkIWroteDown()
      Failed asserting that two strings are equal.
      --- Expected
      +++ Actual
      @@ @@
      -'http://acceptance-app-files/index.php/s/YLCoqRdoz22Mpaa/authenticate/preview'
      +'http://acceptance-app-files/index.php/s/YLCoqRdoz22Mpaa/authenticate/showShare'
    And I see that a wrong password for the shared file message is shown                       # FilesSharingAppContext::iSeeThatAWrongPasswordForTheSharedFileMessageIsShown()

  Scenario: access a direct download shared link protected by password with a valid password                  # /drone/src/github.com/nextcloud/server/tests/acceptance/features/app-files.feature:139
    Given I act as John                                                                                       # ActorContext::iActAs()
    And I am logged in                                                                                        # LoginPageContext::iAmLoggedIn()
    And I share the link for "welcome.txt" protected by the password "abcdef"                                 # FilesAppContext::iShareTheLinkForProtectedByThePassword()
    And I write down the shared link                                                                          # FilesAppContext::iWriteDownTheSharedLink()
    When I act as Jane                                                                                        # ActorContext::iActAs()
    And I visit the direct download shared link I wrote down                                                  # FilesSharingAppContext::iVisitTheDirectDownloadSharedLinkIWroteDown()
    And I see that the current page is the Authenticate page for the direct download shared link I wrote down # FilesSharingAppContext::iSeeThatTheCurrentPageIsTheAuthenticatePageForTheDirectDownloadSharedLinkIWroteDown()
      Failed asserting that two strings are equal.
      --- Expected
      +++ Actual
      @@ @@
      -'http://acceptance-app-files/index.php/s/o4xcQKBSbasioTm/authenticate/download'
      +'http://acceptance-app-files/index.php/s/o4xcQKBSbasioTm/authenticate/downloadShare'
    And I authenticate with password "abcdef"                                                                 # FilesSharingAppContext::iAuthenticateWithPassword()
    And I see that the current page is the Authenticate page for the direct download shared link I wrote down # FilesSharingAppContext::iSeeThatTheCurrentPageIsTheAuthenticatePageForTheDirectDownloadSharedLinkIWroteDown()
```

I do not know, though, if the tests need to be updated to the new URLs or the code should be fixed to keep the old ones.

rullzer committed:
Fix direct access to authen page

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
@MorrisJobke yes that happens because you go directly to the auth page. So there is no state stored yet. Anyways let me fix that.

rullzer commented:
@danxuliu ah you are right. Yes the tests need to be updated.

rullzer committed:
Fix acceptance test

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
All happy!

MorrisJobke reviewed (approved):
Tested and works 👍

juliushaertl commented on apps/files_sharing/js/public.js:
```diff
@@ -223,8 +222,8 @@ OCA.Sharing.PublicApp = {
-				urlSpec.t = $('#dirToken').val();
-				return OC.generateUrl('/apps/files_sharing/ajax/publicpreview.php?') + $.param(urlSpec);
+				var token = $('#dirToken').val();
+				return OC.linkTo('files_sharing', '/publicpreview/'+token) + '?' + OC.buildQueryString(urlSpec);
```
Should be OC.generateUrl(OC.linkTo(...)) otherwise index.php prefix isn't added.

juliushaertl commented on apps/files_sharing/js/public.js:
```diff
@@ -150,15 +149,15 @@ OCA.Sharing.PublicApp = {
 		}
 		else if (previewSupported === 'true') {
-			$('#imgframe > video').attr('poster', OC.filePath('files_sharing', 'ajax', 'publicpreview.php') + '?' + OC.buildQueryString(params));
+			$('#imgframe > video').attr('poster', OC.linkTo('files_sharing', '/publicpreview/'+token) + '?' + OC.buildQueryString(params));
```
Should be OC.generateUrl(OC.linkTo(...)) otherwise index.php prefix isn't added.

juliusknorr reviewed (dismissed):
Small issue with public previews on my instance, but besides that :+1:

rullzer committed:
Fix index.php links

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>
